### ISSUE TITLE
rOpenSci review 1: Fuzzy set math explanation

### ISSUE THREAD

**Issue by llrs** [llrs/BaseSet#39]
**Fuzzy set math explanation**

Consider the following borrowed from the `fuzzy.Rmd` vignette:

```
> set.seed(4567)
> relations <- data.frame(sets = c(rep("A", 5), "B", "C"),
                          elements = c(letters[seq_len(6)], letters[6]),
                          fuzzy = runif(7))
> fuzzy_set <- tidySet(relations)
> fuzzy_set
  elements sets     fuzzy
1        a    A 0.2309186
2        b    A 0.7412554
3        c    A 0.1833834
4        d    A 0.4221682
5        e    A 0.5996399
6        f    B 0.2773313
7        f    C 0.7307312
> BaseSet::union(fuzzy_set, sets=c("C", "B"))
  elements sets     fuzzy
1        f  C∪B 0.7307312
```

When I first saw this result, I thought the math was simply wrong. I would expect that the _probability_ that `f` is in (C or B) would be greater than or equal to the probability that `f` is only in C. Why is it _exactly_ equal? Statistically, this would only be true if B were a subset of C.

After a bit of research, though, I found you are indeed using the standard operator definitions for fuzzy sets. If the fuzziness were the same as probability, then the union operator is a function that returns the lower bound of the probability of membership in the larger set. Do sets of genes expressed in different tissues follow the assumptions the union operator is making?

Fuzzy sets are based on [fuzzy logic](https://plato.stanford.edu/entries/logic-fuzzy). In fuzzy logic, the 0 to 1 values are not probabilities, they are _truth values_. They are measures of the vagueness of a term not the than the chance of membership. You should describe the difference both in your README and the fuzzy vignette.

Most statisticians and bioinformaticians will not be familiar with fuzzy logic. Also, most programs in the field deal with classification probabilities not fuzzy truth values. Although, it is possible that the values associated with tissue annotations are measures of the degree to which a gene is active in the specific tissue, in which case the fuzzy logic may apply.

So, you definitely need to describe the math and statistics behind the operators. Is there any prior research on the application of fuzzy sets in bioinformatics? If so, you might want to add some citations. You also need to justify the choice to use fuzzy set theory rather than more conventional statistics.

**Comment by llrs** [comment]
In this case both sets B and C are 1 element size, both of them have the `f` element. So taking the higher bound of the common values makes the probability of `f` equal to C. Note here that the union does not calculate the probability that `f` is on C or B, but which is the probability of the common elements of C and B. To know the probability of `f` to be in C or B we need to use:

```r
element_size(fuzzy_set, "f")
#   elements size probability
# 6        f    0   0.1945921
# 7        f    1   0.6027532
# 8        f    2   0.2026547
``` 

Here we can see that there is a 0.6 probability that f is in either of them. (If `f` would be in other sets we need to first filter to only those: `fuzzy_set %>% filter(sets %in% c("B", "C")) %>% element_size(element = "f")`, in this case the result should be the same.)

The  standard operator definitions for fuzzy are defined in the `union` argument fun, which is set to "max". However, then the definition of the common operation is not the same as it is implemented as you define it as "lower bound of the probability of membership in the larger set". In this case where both of them have the same number of elements involved, which should be picked? This could be a corner case, but as you explained they must be well covered and documented. In addition to that how is the union the "lower bound of the probability", wouldn't it be the higher bound of the probability? 

I'm not sure I understand the assumptions of the union operator. But the point of having this package is to test if fuzzy methods are helpful on sets and pathways tests. I do not know before hand if the fuzzy logic can explain better any system than other methodologies.

The theoretical background of fuzzy logic is broad and beyond my current comprehension. However, on the linked page there is an example (section 9) on how such a systems arise that is pertinent: "Voting semantics is based on the idea that different agents (voters) may coherently judge the same proposition differently. The proportion of agents that accept a proposition φ as true may be seen as a truth value. Without further restrictions this does not lead to a truth functional semantics, but rather to an assignment of probabilities to propositions. But if one assigns a fixed level of skepticism to each agent and imposes some natural conditions that keep the judgments on logically complex statements consistent with those levels, then one can recover min, max, and 1−x as truth functions ". On the case of pathways and genes there are several voters (scientists or algorithms) that provide their truth value of the association of a gene in a pathway. This is usually expressed as probabilities "I'm 80% sure that X gene is on Y pathway". This is why I don't make a distinction between vagueness and the probability of an association.

I understand the lack of exposure of this field on the bioinformatics, statistics and experimental fields with this logic. This lack of understanding can slow the adoption of the package but at the same time is one of the strengths as it provides with a new method not used much previously on the field. There are other experimental methods to measure the  degree to which a gene is active, however then the problem is distinguish between tissues and between cell types. I build a webpage where I collected an example of the application of BaseSet on this kind of research: https://llrs.github.io/BaseSet_scRNAseq/AUCell.html

There are a couple of articles where fuzzy set methods are used, specially on the area of ontology assignment and comparison. I will provide them as examples of previous usage of these methods. However I haven't cited them because I find them hard to follow and I'm not sure I understand well how they apply the fuzzy logic on their respective topics.

**Comment by llrs** [comment]
```
set.seed(79321)
shape <- sample(c("square", "triangle", "circle", "hexagon"), 70, replace = TRUE)
col <- sample(c("blue", "red", "green", "white", "black"), 70, replace = TRUE)
sets <- sample(c("A", "B", "C"), 70, replace = TRUE)
rel <- data.frame(elements = paste0("element", 1:70), sets = sets, shape = shape, color = col)
TS <- tidySet(rel) %>%
    move_to("relations", "elements", "shape") %>%
    move_to("relations", "elements", "color") %>% 
    group(color == "blue", name = "blue") %>% 
    group(color == "red", name = "red") %>% 
    group(color == "white", name = "white") %>% 
    group(color == "black", name = "black") %>% 
    group(shape == "square", name = "square") %>% 
    group(shape == "triangle", name = "triangle") %>% 
    group(shape == "circle", name = "circle") %>% 
    group(shape == "hexagon", name = "hexagon")

set_size(TS, set = c("A", "B"))

TS %>% 
    union(c("A", "B")) %>% 
    set_size()
set_size(TS, set = c("circle", "white"))

TS %>% 
    union(c("circle", "white")) %>% 
    set_size()

TS %>% 
    filter_element(shape == "circle" & color == "white") %>% 
    elements()

TS2 <- rel %>% 
    cbind(fuzzy = runif(70)) %>% 
    tidySet() %>% 
    move_to("relations", "elements", "shape") %>%
    move_to("relations", "elements", "color") %>% 
    group(color == "blue", name = "blue") %>% 
    group(color == "red", name = "red") %>% 
    group(color == "white", name = "white") %>% 
    group(color == "black", name = "black") %>% 
    group(shape == "square", name = "square") %>% 
    group(shape == "triangle", name = "triangle") %>% 
    group(shape == "circle", name = "circle") %>% 
    group(shape == "hexagon", name = "hexagon")

set_size(TS2, set = c("A", "B"))

TS2 %>% 
    union(c("A", "B")) %>% 
    set_size()
set_size(TS2, set = c("circle", "white"))

TS2 %>% 
    union(c("circle", "white")) %>% 
    set_size()

filter_element(TS2, shape == "circle" & color == "white")
```

**Comment by llrs** [comment]
Reviwer 2;

 - [x] Vignette(s) demonstrating major functionality that runs successfully locally
 - [x] Examples for all exported functions in R Help that run successfully locally
 - [x] Functionality: Any functional claims of the software been confirmed.

 - Pluralized some arguments of set_size and element_size to match the expected name of columns
 - Modified the examples so that sets are represented by upper case letters and elements by lower case letters.
 - About the weird behaviour of `set_size()`
Your first try to modify the set and convert to a fuzzy set you'll need to assign the modified ecoli_set to be able to use the fuzzy values (or pipe it). The TidySet is of class S4, not S6, and the modifications are not in place. 

```r
ecoli_sets %>%
      mutate(
      fuzzy = case_when(sets == "GL" ~ 0.2,    # Add fuzzy values
      sets == "CF" ~ 0.8)) %>%
      set_size()
```

Yes, the GL has 0.04 probability to have two genes so the glycolysis pathway probably don't have any genes. I think this is different from cardinality, as cardinality it is just a number for a set, while here I return several values for a single set. I added to the documentation how is this calculated (via `length_set`).

I'm not sure I follow on this cardinality. For example the `{sets}` package mentioned provides cardinality as the number of elements in a set regardless of the fuzzy value. However I will provide a new method to calculate cardinality and allow to provide a fuzzy logic function (`length` as in `{sets}` or `sum` as you found, or other functions).  

 - [x] Vignette comments

Changed the automatic names for the full names, also added some links for references to the data and how to interpret it. I tried to use some fuzzy values, but it would take too long to calculate for many sets. 

----

**Fuzzy logic & explanation**

I modified the README section about fuzzy sets, now it reads : "Fuzzy sets are similar to classical sets but there is some vagueness on the relationship between the element and the set.".

Similarly I've added a new section on the README about why this package was developed and why it could be useful to others. Also introduced the topic on the fuzzy vignette. 

I do not want to force an interpretation of the fuzzy theory which sometimes translate to using different semantics: membership, probabilities, truth vales, and others. I think that the fuzzy column is the most neutral and better understood. There seems to be a movement from fuzzy research to [differentiate it from probabilities](http://www.cs.cmu.edu/Groups/AI/html/faqs/ai/fuzzy/part1/faq-doc-10.html), but at the same time to [expand them](https://en.wikipedia.org/wiki/Fuzzy_logic#Comparison_to_probability)(probabilities). 

One can use the logic you desire according to the specific framework one wish to work. This in encouraged through the package by being able to use any logic.To show the flexibility of the package using the same data I've added some examples with different logic, explaining when could this logic be applied. Hope this way it is clear enough that the package does not follow a certain logic, but that the defaults are from the standard fuzzy logic. 

On the specific case from the vignette I added some sentences explaining why I am using it. `union(TS, sets = c("A", "B"), FUN = function(x){y <- sum(x)-prod(x); ifelse(length(x)==1, x, y)})`.
